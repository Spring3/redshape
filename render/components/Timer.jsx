import React, { Fragment, Component } from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import PlayIcon from 'mdi-react/PlayIcon';
import PauseIcon from 'mdi-react/PauseIcon';
import StopIcon from 'mdi-react/StopIcon';

import actions from '../actions';

import { GhostButton } from './Button';
import { animationSlideUp } from '../animations';
import Link from './Link';

import { setupTimerIPC, sendTimerIPC } from '../ipc';

const ActiveTimer = styled.div`
  animation: ${animationSlideUp} .7s ease-in;
  max-width: 100%;
  box-sizing: border-box;
  padding: 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  background: ${props => props.theme.bg};
  display: ${props => props.isEnabled ? 'flex' : 'none'};
  align-items: center;
  box-shadow: 0px -2px 20px ${props => props.theme.bgDark};
  border-top: 2px solid ${props => props.theme.bgDark};

  div.buttons {
    margin: 0 20px;
  }

  div.time {
    margin: 0 20px;
    font-size: 16px;
    font-weight: bold;
  }

  div.buttons {
    a {
      padding: 10px 0px;

      &:hover {
        background: ${props => props.theme.bgDark};
      }
    }

    a:first-child {
      margin-right: 20px;
    }
  }

  div.time {
    color: ${props => props.theme.main};
  }
`;

const StyledButton = styled(GhostButton)`
  padding: 0px;
`;

const MaskedLink = styled(Link)`
  color: inherit;
  padding: 0;
  margin: 0 20px;
  font-size: 16px;
  font-weight: bold;
  text-decoration: none;
`;

class Timer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: props.trackedTime || props.initialValue || 0,
      timestamp: null, // used only when store/restore to/from timestamp (window hidden)
    };

    setupTimerIPC(this);

    const { isEnabled, isPaused, trackedIssue } = this.props;
    if (isEnabled){
      sendTimerIPC('timer-info', {isEnabled, isPaused, issue: trackedIssue})
    }

    this.interval = props.isEnabled && !props.isPaused
     ? setInterval(this.tick, 1000)
     : undefined;
  }
  /* stop time interval and store tracked time + current datetime */
  storeToTimestamp(){
    const { timestamp } = this.state;
    const { isEnabled, isPaused, trackedIssue } = this.props;
    if (isEnabled && !isPaused){
      this.setState({
        timestamp: {
          value: this.state.value,
          datetime: moment()
        }
      });
      this.stopInterval()
    }
    sendTimerIPC('timer-info', {isEnabled, isPaused, issue: trackedIssue})
  }
  /* restore time interval based on stored tracked time + diff datetimes */
  restoreFromTimestamp(enableInterval = true){
    const { timestamp } = this.state;
    const { isEnabled, isPaused } = this.props;
    if (timestamp && isEnabled && !isPaused){
      const { value, datetime } = timestamp;
      const newValue = moment().diff(datetime, 'ms') + value;
      this.setState({ timestamp: null, value: newValue });
      if (enableInterval) {
        this.interval = setInterval(this.tick, 1000);
      }
    }
  }

  tick = () => {
    const { value } = this.state;
    this.setState({ value: value + 1000 });
  }

  stopInterval = () => {
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = undefined;
    }
  }

  cleanup = () => {
    const { isEnabled, isPaused } = this.props;
    if (isEnabled && !isPaused) {
      this.onPause();
    }
    this.stopInterval();
    window.removeEventListener('beforeunload', this.cleanup);
  }

  saveState = () => {
    const { isEnabled, isPaused } = this.props;
    if (isEnabled && !isPaused) {
      const { continueTimer } = this.props;
      continueTimer(this.state.value);
    }
    this.stopInterval();
  }

  componentWillMount() {
    window.addEventListener('beforeunload', this.cleanup);
  }

  componentWillUnmount() {
    this.saveState();
  }

  componentDidUpdate(oldProps) {
    const { isEnabled, isPaused, trackedIssue } = this.props;
    if (isEnabled !== oldProps.isEnabled) {
      this.stopInterval();
      // if was disabled, but now is enabled
      if (isEnabled) {
        this.interval = setInterval(this.tick, 1000);
        sendTimerIPC('timer-info', {isEnabled, isPaused, issue: trackedIssue });
      }
      // otherwise, if was enabled, but now it's disabled, we don't do anything
      // because the interval was already cleared above
    }
  }

  onPause = () => {
    this.stopInterval();
    const { onPause, trackedIssue, pauseTimer } = this.props;
    const { value } = this.state;
    pauseTimer(value);
    if (onPause) {
      onPause(value, trackedIssue)
    }
    sendTimerIPC('timer-info', {isEnabled: true, isPaused: true, issue: trackedIssue})
  }

  onContinue = () => {
    this.interval = setInterval(this.tick, 1000);
    const { onContinue, trackedIssue, continueTimer } = this.props;
    const { value } = this.state;
    continueTimer(value);
    if (onContinue) {
      onContinue(trackedIssue, value);
    }
    sendTimerIPC('timer-info', {isEnabled: true, isPaused: false, issue: trackedIssue})
  }

  onStop = () => {
    this.stopInterval();
    const { value } = this.state;
    const { onStop, trackedIssue, stopTimer } = this.props;
    stopTimer(value);
    if (onStop) {
      onStop(value, trackedIssue);
    }
    this.setState({ value: 0 });
    sendTimerIPC('timer-info', {isEnabled: false, issue: trackedIssue})
  }

  redirectToTrackedLink = (event) => {
    event.preventDefault();
    this.props.history.push(`/app/issue/${this.props.trackedIssue.id}`);
  }

  componentWillReceiveProps(newProps) {
    const { trackedTime } = newProps;
    this.setState({ value: trackedTime })
  }

  render() {
    const { value } = this.state;
    const { isEnabled, trackedIssue, isPaused } = this.props;
    const timeString = moment.utc(value).format('HH:mm:ss');
    return (
      <Fragment>
      <ActiveTimer isEnabled={isEnabled}>
        <div className="buttons">
          <StyledButton
            onClick={this.onStop}
          >
            <StopIcon size={35} />
          </StyledButton>
          { isPaused && (
              <StyledButton
                onClick={this.onContinue}
              >
                <PlayIcon size={35} />
              </StyledButton>
            )
          }
          { !isPaused && (
              <StyledButton
                onClick={this.onPause}
              >
                <PauseIcon size={35} />
              </StyledButton>
            )
          }
        </div>
        <div className="issueName">
          { (isEnabled ?
          <MaskedLink
            href='#'
            onClick={this.redirectToTrackedLink}
          >
            {trackedIssue.subject}
          </MaskedLink>
            : null)}
        </div>
        <div className="time">
          <span>{timeString}</span>
        </div>
      </ActiveTimer>
      </Fragment>
    );
  }
}

Timer.propTypes = {
  isEnabled: PropTypes.bool,
  isPaused: PropTypes.bool,
  initialValue: PropTypes.number,
  trackedIssue: PropTypes.shape({
    id: PropTypes.number.isRequired,
    subject: PropTypes.string.isRequired,
    author: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    project: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    activity: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  trackedTime: PropTypes.number,
  onStop: PropTypes.func,
  onPause: PropTypes.func,
  onContinue: PropTypes.func,
  pauseTimer: PropTypes.func.isRequired,
  continueTimer: PropTypes.func.isRequired,
  stopTimer: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired
};

Timer.defaultProps = {
  initialValue: 0,
  isEnabled: false,
  isPaused: false,
  issueTitle: ''
};

const mapStateToProps = state => ({
  isEnabled: state.tracking.isEnabled,
  isPaused: state.tracking.isPaused,
  trackedTime: state.tracking.duration,
  trackedIssue: state.tracking.issue,
});

const mapDispatchToProps = dispatch => ({
  pauseTimer: value => dispatch(actions.tracking.trackingPause(value)),
  continueTimer: value => dispatch(actions.tracking.trackingContinue(value)),
  stopTimer: value => dispatch(actions.tracking.trackingStop(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(Timer);
