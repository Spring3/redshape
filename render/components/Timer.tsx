import React, {
  useState, useEffect, useRef, useCallback
} from 'react';
// eslint-disable-next-line
import moment from 'moment';
import styled from 'styled-components';
import PlayIcon from 'mdi-react/PlayIcon';
import PauseIcon from 'mdi-react/PauseIcon';
import StopIcon from 'mdi-react/StopIcon';

import { GhostButton } from './GhostButton';
import { animationSlideUp } from '../animations';
import { Link } from './Link';
import { useOvermindActions, useOvermindState } from '../store';
import { Issue, TimeTrackingAction } from '../../types';

const ActiveTimer = styled.div`
  animation: ${animationSlideUp} .7s ease-in;
  max-width: 100%;
  box-sizing: border-box;
  padding: 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  background: ${(props) => props.theme.bg};
  display: ${(props) => (props.isEnabled ? 'flex' : 'none')};
  align-items: center;
  box-shadow: 0px -2px 20px ${(props) => props.theme.bgDark};
  border-top: 2px solid ${(props) => props.theme.bgDark};
  
  div.panel {
    flex-grow: 0;
    min-width: 520px;
    display: flex;
    align-items: center;
    max-width: ${(props) => (props.advancedTimerControls ? '900px' : '1800px')};
  }
  
  div.buttons {
    margin: 0 20px;
    display: flex;
  }
  
  div.time {
    margin: 0 20px;
    font-size: 16px;
    font-weight: bold;
  }

  div.buttons {
    a {
      padding: 10px 0px;

      &:hover {
        background: ${(props) => props.theme.bgDark};
      }
    }

    a:first-child {
      margin-right: 20px;
    }
  }
  
  div.buttons.buttons-advanced {
    a {
      margin-right: 5px;
    }
    a:last-child {
      margin-right: initial;
    }
  }
  
  div.issueName {
    padding: 0 20px;
    max-width: 500px;
  }

  div.time {
    color: ${(props) => props.theme.main};
  }
  
  input[name="comment"] {
    flex-grow: 2;
    margin-left: 20px;
    width: initial;
    border: none;
    border-radius: 0;
    border-bottom: 1px ${(props) => props.theme.bgDark} solid;
    color: #A4A4A4;
    &:focus {
      border: none;
      border-radius: 0;
      border-bottom: 1px ${(props) => props.theme.main} solid;
      box-shadow: none;
    }
  }
  
`;

const StyledButton = styled(GhostButton)`
  padding: 0px;
`;

const MaskedLink = styled(Link)`
  color: inherit;
  padding: 0;
  font-size: 16px;
  font-weight: bold;
  text-decoration: none;
`;

type TimerEventHook = (args: { issue: Issue, recordedTime: number }) => void;

type TimerProps = {
  startFrom?: number;
  autoStart?: boolean;
  onStop?: TimerEventHook;
  onStart?: TimerEventHook;
  onPause?: TimerEventHook;
  onContinue?: TimerEventHook;
}

const Timer = ({ startFrom = 0, autoStart = false, onStop, onStart, onPause, onContinue }: TimerProps) => {
  const [timeMs, setTimeMs] = useState(startFrom);
  const interval = useRef<ReturnType<typeof setInterval>>();
  const state = useOvermindState();
  const actions = useOvermindActions();

  const [lastTimeRecord] = state.timeTracking.records.slice(-1);
  const issueId = lastTimeRecord?.issueId;
  const targetIssue = state.issues.byId[issueId];

  const isCounting = [TimeTrackingAction.CONTINUE, TimeTrackingAction.START].includes(lastTimeRecord?.action);
  const isPaused = lastTimeRecord?.action === TimeTrackingAction.PAUSE;
  const isStopped = lastTimeRecord?.action === TimeTrackingAction.STOP;

  const tick = useCallback(() => {
    setTimeMs((trackedTimeMs) => trackedTimeMs + 1000);
  }, []);

  const start = useCallback(() => {
    if (isPaused || isStopped) {
      actions.timeTracking.track({ issueId });
      if (!interval.current) {
        interval.current = setInterval(tick, 1000);
      }
      if (onStart) {
        onStart({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.track, issueId, tick, targetIssue, onStart, timeMs]);

  const stop = useCallback(() => {
    if (isCounting) {
      actions.timeTracking.stop();
      if (interval.current) {
        clearInterval(interval.current);
      }
      if (onStop) {
        onStop({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.stop, onStop, targetIssue, timeMs]);

  const pause = useCallback(() => {
    if (isCounting) {
      actions.timeTracking.pause();
      if (interval.current) {
        clearInterval(interval.current);
      }
      if (onPause) {
        onPause({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.pause, onPause, targetIssue, timeMs]);

  const unpaunse = useCallback(() => {
    if (isPaused) {
      actions.timeTracking.unpause();
      if (!interval.current) {
        interval.current = setInterval(tick, 1000);
      }
      if (onContinue) {
        onContinue({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.unpause, tick, onContinue, targetIssue, timeMs]);

  useEffect(() => {
    if (isCounting) {
      setTimeMs(startFrom);
      if (autoStart) {
        start();
      }
    }
  }, [startFrom, autoStart, start]);

  const timeString = moment.utc(timeMs).format('HH:mm:ss');
  return (
    <>
      <ActiveTimer isEnabled={lastTimeRecord && !isStopped}>
        <div className="panel">
          <div className="buttons">
            <StyledButton
              id="stop-timer"
              onClick={stop}
            >
              <StopIcon size={35} />
            </StyledButton>
            { isPaused
              ? (
                <StyledButton
                  id="continue-timer"
                  onClick={unpaunse}
                >
                  <PlayIcon size={35} />
                </StyledButton>
              )
              : (
                <StyledButton
                  id="pause-timer"
                  onClick={pause}
                >
                  <PauseIcon size={35} />
                </StyledButton>
              )}
          </div>
          <div className="issueName">
            { !isStopped
              ? (
                <MaskedLink
                  href="#"
                >
                  {targetIssue?.subject}
                </MaskedLink>
              )
              : null}
          </div>
          <div className="time">
            <span>{timeString}</span>
          </div>
        </div>
      </ActiveTimer>
    </>
  );
};

export {
  Timer
};
