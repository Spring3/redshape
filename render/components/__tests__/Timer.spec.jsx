import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

// using mount instead of shallow due to https://github.com/airbnb/enzyme/issues/1908

import { Timer } from '../Timer';

const mockStore = configureStore([thunk]);

const stateSettings = {
  idleBehavior: 0,
  discardIdleTime: true,
  advancedTimerControls: false,
  progressWithStep1: false,
};

const waitSeconds = (n = 1) => new Promise((resolve) => setTimeout(() => resolve(), n * 1000));

describe('Timer component', () => {
  afterEach(cleanup);
  it('should automatically resume timer if trackedTime prop was provided', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 4000,
        comments: '',
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      },
      settings: stateSettings,
    };

    const store = mockStore(state);

    render(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    await waitSeconds(1);
    const buttons = document.querySelector('.buttons');
    expect(buttons.children).toHaveLength(2);
    expect(buttons.lastElementChild).toHaveAttribute('id', 'pause-timer');
    fireEvent.click(buttons.lastElementChild);
    expect(onPause).toHaveBeenCalled();
  });

  it('should start the counter from the initialValue if given', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      },
      settings: stateSettings,
    };

    const store = mockStore(state);
    const initialValue = 3000;

    const { unmount } = render(
      <Provider store={store}>
        <Timer
          initialValue={initialValue}
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    await waitSeconds(1);
    fireEvent.click(document.querySelector('.buttons').lastElementChild);
    unmount();
  });

  it('should automatically save the progress before unmount', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 4000,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      },
      settings: stateSettings,
    };

    const store = mockStore(state);

    const { unmount } = render(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    await waitSeconds(1);
    unmount();

    expect(onPause).toHaveBeenCalled();
    expect(store.getActions()).toEqual([{ type: 'TRACKING_PAUSE', data: { duration: 5000, comments: '' } }]);
  });

  it('should allow to resume the paused timer', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: true,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      },
      settings: stateSettings,
    };

    const store = mockStore(state);

    const { unmount } = render(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const buttons = document.querySelector('.buttons');
    expect(buttons).toBeDefined();
    expect(buttons.children).toHaveLength(2);
    expect(buttons.lastElementChild).toHaveAttribute('id', 'continue-timer');

    fireEvent.click(buttons.lastElementChild);
    await waitSeconds(1);
    expect(onContinue).toHaveBeenCalledTimes(1);
    unmount();
    expect(store.getActions().pop()).toEqual({ type: 'TRACKING_SAVE', data: { duration: 1000, comments: '' } });
  });

  it('should be able to stop time timer', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      },
      settings: stateSettings,
    };

    const store = mockStore(state);

    render(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const buttons = document.querySelector('.buttons');
    expect(buttons).toBeDefined();
    expect(buttons.children).toHaveLength(2);
    expect(buttons.firstElementChild).toHaveAttribute('id', 'stop-timer');

    await waitSeconds(1);
    fireEvent.click(buttons.firstElementChild);

    expect(onStop).toHaveBeenCalledWith(state.tracking.issue, 1000, '');
    expect(onStop).toHaveBeenCalledTimes(1);
  });

  it('should redirect when the name of the tracked issue is clicked', () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: true,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      },
      settings: stateSettings,
    };

    const store = mockStore(state);
    const history = {
      push: jest.fn()
    };

    render(
      <Provider store={store}>
        <Timer
          history={history}
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    fireEvent.click(document.querySelector('.issueName').firstElementChild);
    expect(history.push).toHaveBeenCalledWith(`/app/issue/${state.tracking.issue.id}`);
  });
});
