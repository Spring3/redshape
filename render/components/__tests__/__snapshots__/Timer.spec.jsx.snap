// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer component should match the snapshot 1`] = `
.c2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: color ease;
  transition: color ease;
}

.c2:active,
.c2:focus,
.c2:visited {
  background: transparent;
}

.c4 {
  font-size: 14px;
  padding: 2px;
  cursor: pointer;
}

.c4:hover {
  background: undefined;
  color: undefined;
}

.c0 {
  -webkit-animation: gOHwoP .7s ease-in;
  animation: gOHwoP .7s ease-in;
  max-width: 100%;
  box-sizing: border-box;
  padding: 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0px -2px 20px;
  border-top: 2px solid;
}

.c0 div.panel {
  min-width: 520px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 div.buttons {
  margin: 0 20px;
}

.c0 div.time {
  margin: 0 20px;
  font-size: 16px;
  font-weight: bold;
}

.c0 div.buttons a {
  padding: 10px 0px;
}

.c0 div.buttons a:first-child {
  margin-right: 20px;
}

.c0 input[name="comment"] {
  border: none;
  border-radius: 0;
  border-bottom: 1px solid;
  color: #A4A4A4;
}

.c0 input[name="comment"]:focus {
  border: none;
  border-radius: 0;
  border-bottom: 1px solid;
  box-shadow: none;
}

.c1 {
  padding: 0px;
}

.c3 {
  color: inherit;
  padding: 0;
  margin: 0 20px;
  font-size: 16px;
  font-weight: bold;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Timer)
    trackedDuration={4000}
  >
    <Timer
      advancedTimerControls={false}
      continueTimer={[Function]}
      discardIdleTime={true}
      idleBehavior={0}
      initialValue={0}
      isEnabled={true}
      isPaused={true}
      issueTitle=""
      pauseTimer={[Function]}
      saveTimer={[Function]}
      stopTimer={[Function]}
      trackedDuration={4000}
      trackedIssue={
        Object {
          "id": "123abc",
          "subject": "Test issue",
        }
      }
      trackedTime={0}
    >
      <Timer__ActiveTimer
        isEnabled={true}
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Timer__ActiveTimer-d99dh7-0",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "animation:",
                  Keyframes {
                    "id": "sc-keyframes-gOHwoP",
                    "inject": [Function],
                    "name": "gOHwoP",
                    "rules": Array [
                      "@-webkit-keyframes gOHwoP{0%{opacity:0;-webkit-transform:translateY(100px);-ms-transform:translateY(100px);transform:translateY(100px);}100%{opacity:1;-webkit-transform:translateY(0px);-ms-transform:translateY(0px);transform:translateY(0px);}}",
                      "@keyframes gOHwoP{0%{opacity:0;-webkit-transform:translateY(100px);-ms-transform:translateY(100px);transform:translateY(100px);}100%{opacity:1;-webkit-transform:translateY(0px);-ms-transform:translateY(0px);transform:translateY(0px);}}",
                    ],
                    "toString": [Function],
                  },
                  " .7s ease-in;max-width:100%;box-sizing:border-box;padding:20px;position:fixed;bottom:0;width:100%;background:",
                  [Function],
                  ";display:",
                  [Function],
                  ";align-items:center;box-shadow:0px -2px 20px ",
                  [Function],
                  ";border-top:2px solid ",
                  [Function],
                  ";div.panel{min-width:520px;display:flex;align-items:center;}div.buttons{margin:0 20px;}div.time{margin:0 20px;font-size:16px;font-weight:bold;}div.buttons{a{padding:10px 0px;&:hover{background:",
                  [Function],
                  ";}}a:first-child{margin-right:20px;}}div.time{color:",
                  [Function],
                  ";}input[name=\\"comment\\"]{border:none;border-radius:0;border-bottom:1px ",
                  [Function],
                  " solid;color:#A4A4A4;&:focus{border:none;border-radius:0;border-bottom:1px ",
                  [Function],
                  " solid;box-shadow:none;}}",
                ],
              },
              "displayName": "Timer__ActiveTimer",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Timer__ActiveTimer-d99dh7-0",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          isEnabled={true}
        >
          <div
            className="c0"
          >
            <div
              className="panel"
            >
              <div
                className="buttons"
              >
                <Timer__StyledButton
                  onClick={[Function]}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "$asyncbind": [Function],
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "Timer__StyledButton-d99dh7-1",
                          "isStatic": true,
                          "lastClassName": "c1",
                          "rules": Array [
                            "padding:0px;",
                          ],
                        },
                        "displayName": "Timer__StyledButton",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "Timer__StyledButton-d99dh7-1",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[Function]}
                  >
                    <GhostButton
                      className="c1"
                      disabled={false}
                      onClick={[Function]}
                    >
                      <Button__StyledLink
                        className="c1"
                        disabled={false}
                        href="#"
                        onClick={[Function]}
                      >
                        <StyledComponent
                          className="c1"
                          disabled={false}
                          forwardedComponent={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": Array [],
                              "componentStyle": ComponentStyle {
                                "componentId": "Button__StyledLink-bgocaa-1",
                                "isStatic": false,
                                "lastClassName": "c2",
                                "rules": Array [
                                  "text-decoration:none;color:",
                                  [Function],
                                  ";transition:color ease ",
                                  [Function],
                                  ";",
                                  [Function],
                                  " &:active,&:focus,&:visited{background:transparent;}",
                                ],
                              },
                              "displayName": "Button__StyledLink",
                              "foldedComponentIds": Array [],
                              "render": [Function],
                              "styledComponentId": "Button__StyledLink-bgocaa-1",
                              "target": "a",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                          href="#"
                          onClick={[Function]}
                        >
                          <a
                            className="c1 c2"
                            disabled={false}
                            href="#"
                            onClick={[Function]}
                          >
                            <StopIcon
                              size={35}
                            >
                              <path
                                d="M18,18H6V6H18V18Z"
                              />
                            </StopIcon>
                          </a>
                        </StyledComponent>
                      </Button__StyledLink>
                    </GhostButton>
                  </StyledComponent>
                </Timer__StyledButton>
                <Timer__StyledButton
                  onClick={[Function]}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "$asyncbind": [Function],
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "Timer__StyledButton-d99dh7-1",
                          "isStatic": true,
                          "lastClassName": "c1",
                          "rules": Array [
                            "padding:0px;",
                          ],
                        },
                        "displayName": "Timer__StyledButton",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "Timer__StyledButton-d99dh7-1",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[Function]}
                  >
                    <GhostButton
                      className="c1"
                      disabled={false}
                      onClick={[Function]}
                    >
                      <Button__StyledLink
                        className="c1"
                        disabled={false}
                        href="#"
                        onClick={[Function]}
                      >
                        <StyledComponent
                          className="c1"
                          disabled={false}
                          forwardedComponent={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": Array [],
                              "componentStyle": ComponentStyle {
                                "componentId": "Button__StyledLink-bgocaa-1",
                                "isStatic": false,
                                "lastClassName": "c2",
                                "rules": Array [
                                  "text-decoration:none;color:",
                                  [Function],
                                  ";transition:color ease ",
                                  [Function],
                                  ";",
                                  [Function],
                                  " &:active,&:focus,&:visited{background:transparent;}",
                                ],
                              },
                              "displayName": "Button__StyledLink",
                              "foldedComponentIds": Array [],
                              "render": [Function],
                              "styledComponentId": "Button__StyledLink-bgocaa-1",
                              "target": "a",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                          href="#"
                          onClick={[Function]}
                        >
                          <a
                            className="c1 c2"
                            disabled={false}
                            href="#"
                            onClick={[Function]}
                          >
                            <PlayIcon
                              size={35}
                            >
                              <path
                                d="M8,5.14V19.14L19,12.14L8,5.14Z"
                              />
                            </PlayIcon>
                          </a>
                        </StyledComponent>
                      </Button__StyledLink>
                    </GhostButton>
                  </StyledComponent>
                </Timer__StyledButton>
              </div>
              <div
                className="issueName"
              >
                <Timer__MaskedLink
                  href="#"
                  onClick={[Function]}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "$asyncbind": [Function],
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "Timer__MaskedLink-d99dh7-2",
                          "isStatic": true,
                          "lastClassName": "c3",
                          "rules": Array [
                            "color:inherit;padding:0;margin:0 20px;font-size:16px;font-weight:bold;text-decoration:none;",
                          ],
                        },
                        "displayName": "Timer__MaskedLink",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "Timer__MaskedLink-d99dh7-2",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    href="#"
                    onClick={[Function]}
                  >
                    <Link
                      className="c3"
                      href="#"
                      onClick={[Function]}
                    >
                      <Link__StyledLink
                        className="c3"
                        href="#"
                        onClick={[Function]}
                        rel="noopener noreferer"
                      >
                        <StyledComponent
                          className="c3"
                          forwardedComponent={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": Array [],
                              "componentStyle": ComponentStyle {
                                "componentId": "Link__StyledLink-qed6kb-0",
                                "isStatic": false,
                                "lastClassName": "c4",
                                "rules": Array [
                                  "color:",
                                  [Function],
                                  ";font-size:14px;padding:2px;cursor:pointer;",
                                  [Function],
                                ],
                              },
                              "displayName": "Link__StyledLink",
                              "foldedComponentIds": Array [],
                              "render": [Function],
                              "styledComponentId": "Link__StyledLink-qed6kb-0",
                              "target": "a",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                          href="#"
                          onClick={[Function]}
                          rel="noopener noreferer"
                        >
                          <a
                            className="c3 c4"
                            href="#"
                            onClick={[Function]}
                            rel="noopener noreferer"
                          >
                            Test issue
                          </a>
                        </StyledComponent>
                      </Link__StyledLink>
                    </Link>
                  </StyledComponent>
                </Timer__MaskedLink>
              </div>
              <div
                className="time"
              >
                <span>
                  00:00:00
                </span>
              </div>
            </div>
          </div>
        </StyledComponent>
      </Timer__ActiveTimer>
    </Timer>
  </Connect(Timer)>
</Provider>
`;
